Add versioning requirement with operation and constraint matching

novafacing opened the pull request:
This PR adds a new struct `VersioningReq`, an analogue to `VersionReq` from `semver` but using `Versioning`s instead. This makes it easier to use `Versioning` when interpreting dependency version requirements for non-Rust languages and frameworks.

This PR is fairly complete, it includes documentation, tests, doctests, etc. Happy to discuss/change anything if needed.

fosskers commented on src/lib.rs:
```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
+        Self { op, version }
+    }
+
+    /// Return the operation of the requirement.
+    pub fn op(&self) -> Op {
+        self.op
+    }
```
Why not just make the struct fields `pub` and leave these methods out?

novafacing replied:
Sounds good!

fosskers commented on src/lib.rs:
```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
+    /// use versions::{VersioningReq, Versioning};
+    /// use std::str::FromStr;
+    ///
+    /// let constraint_gt = VersioningReq::from_str(">=1.0.0").expect("Invalid req");
+    /// assert!(constraint_gt.matches(&Versioning::new("1.0.0").expect("Invalid version")));
+    /// assert!(constraint_gt.matches(&Versioning::new("1.1.0").expect("Invalid version")));
+    /// assert!(!constraint_gt.matches(&Versioning::new("0.9.0").expect("Invalid version")));
+    ///
+    /// let constraint_wild = VersioningReq::from_str("*").expect("Invalid req");
+    /// assert!(constraint_wild.matches(&Versioning::new("1.0.0").expect("Invalid version")));
+    /// assert!(constraint_wild.matches(&Versioning::new("1.1.0").expect("Invalid version")));
+    /// assert!(constraint_wild.matches(&Versioning::new("0.9.0").expect("Invalid version")));
+    ///
+    /// let constraint_eq = VersioningReq::from_str("==1.0.0").expect("Invalid req");
+    /// assert!(constraint_eq.matches(&Versioning::new("1.0.0").expect("Invalid version")));
+    /// assert!(!constraint_eq.matches(&Versioning::new("1.1.0").expect("Invalid version")));
+    /// assert!(!constraint_eq.matches(&Versioning::new("0.9.0").expect("Invalid version")));
```
You can just `unwrap` in these doctests instead of supplying a failure reason with `expect`. It would be a bit cleaner.

novafacing replied:
Ah, was using forbid unwrap and forgot to change it :)

fosskers commented on src/lib.rs:
```diff
@@ -57,8 +57,9 @@ use nom::combinator::{fail, map, map_res, opt, peek, recognize, value};
 use nom::multi::separated_list1;
 use nom::IResult;
 use parsers::{alphanums, hyphenated_alphanums, unsigned};
+use semver::Op;
```
I would prefer if we implemented this ourselves instead of depending on `semver`.

novafacing replied:
Sounds good, do we want a feature `semver` which will provide `versioning::Op::into::<semver::Op>` for downstream smoothness?

novafacing replied:
Never mind -- thought about it and I don't see a good reason to have it particularly, but it can always be added later :)

fosskers replied:
Yup let's leave it out entirely.

fosskers commented on src/lib.rs:
```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
+            f,
+            "{}{}",
+            // NOTE: Display is not implemented for `Op`
+            match self.op {
+                Op::Exact => "=",
+                Op::Greater => ">",
+                Op::GreaterEq => ">=",
+                Op::Less => "<",
+                Op::LessEq => "<=",
+                Op::Tilde => "~",
+                Op::Caret => "^",
+                Op::Wildcard => "",
+                _ => Err(std::fmt::Error)?,
+            },
+            self.version
+                .as_ref()
+                .map(|v| v.to_string())
+                .unwrap_or_default()
```
These can be factored out as variables for that the `write!` occurs on a single line.

fosskers commented on src/lib.rs:
```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
+
+/// Parse an operation from the beginning of a string, returning the operation
+/// and the remaining string.
+fn op(input: &str) -> Result<(Op, &str), Error> {
+    let comparator: String = input
+        .chars()
+        .take_while(|c| matches!(c, '=' | '>' | '<' | '~' | '^' | '*'))
+        .collect();
+    let rest = &input[comparator.len()..];
+    let comp = match comparator.as_ref() {
+        "==" => Op::Exact,
+        ">" => Op::Greater,
+        ">=" => Op::GreaterEq,
+        "<" => Op::Less,
+        "<=" => Op::LessEq,
+        "~" => Op::Tilde,
+        "^" => Op::Caret,
+        "*" => Op::Wildcard,
+        "" => Op::Exact,
+        _ => return Err(Error::IllegalOp(comparator)),
+    };
+
+    Ok((comp, rest))
+}
```
This library already uses `nom` quite extensively, why not adopt `nom` usage here too?

novafacing replied:
Converted the op parsing to use nom.

fosskers commented on src/lib.rs:
```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
+/// Checks whether two versioning triples are "tilde-compatible", that is v2's patch version
+/// may be greater than v1's, but its major and minor versions may not be.
+/// For tilde matches, the v2 patch can be greater than the v1 patch
+fn version_triples_match_tilde(v1: &Versioning, v2: &Versioning) -> bool {
```
Perhaps this should be a method on `Versioning` directly.

fosskers commented on src/lib.rs:
```diff
@@ -1378,6 +1382,289 @@ impl Default for Versioning {
+
+/// Checks whether two versioning triples are "caret-compatible", that is v2's parts right of the
+/// first non-zero part may be greater than v1's.
+fn version_triples_match_caret(v1: &Versioning, v2: &Versioning) -> bool {
```
This as well.

fosskers commented on src/lib.rs:
```diff
@@ -1656,4 +1943,135 @@ mod tests {
+
+    #[test]
+    fn test_eq() {
+        assert!(VersioningReq::from_str("==1.0.0")
+            .expect("Failed")
+            .matches(&Versioning::new("1.0.0").expect("Failed")));
```
Here too for all of these tests; it's okay to just `unwrap`.

fosskers reviewed (changes requested):
Thanks for this! I like the concept, yet there are a few things I think we can improve.

novafacing commented:
Above changes resolve all the comments, thanks for the suggestions! I also fixed a mistake that allowed `==` instead of just `=`, and removed the permissive "if there is no op, treat it as `=`" behavior because I think it's likely to cause confusion.

fosskers commented:
Thanks, I'm tweaking a few things and then will merge.

fosskers commented:
Take a look at what I altered. I'd like your input about the naming of the `Requirement` type. It could also potentially be called `Constraint`.

novafacing commented:
Looks great, I'm also in favor of `Requirement` as the name. Changes to make the `parse` functions consistent are nice too!

fosskers commented:
Great, then I'll merge this tomorrow morning and release.

fosskers commented:
Released `6.1.0`.
